src/generators/typescript/presets/JsonBinPackPreset.ts

```diff
@@ -0,0 +1,55 @@
+import { TypeScriptPreset } from '../TypeScriptPreset';
+const alterschema = require('alterschema');
+
+function getInputSchema(originalInput: any) {
+  if(originalInput.$schema !== undefined) {
+    if(originalInput.$schema.includes('http://json-schema.org/draft-04/schema')) {
+      return 'draft4';
+    }
+    if(originalInput.$schema.includes('http://json-schema.org/draft-06/schema')) {
+      return 'draft6';
+    }
+  }
+  return 'draft7';
+}
+
+/**
+ * Preset which adds descriptions
+ *
+ * @implements {TypeScriptPreset}
+ */
+export const TS_JSONBINPACK_PRESET: TypeScriptPreset = {
+  class: {
+    async additionalContent({renderer, content, model}) {
+      renderer.addDependency('const jsonbinpack = require(\'jsonbinpack\');');
+      const jsonSchema = await alterschema(model.originalInput, getInputSchema(model.originalInput), '2020-12');
+      const json = JSON.stringify(jsonSchema);
+      const packContent = `
+public async jsonbinSerialize(): Promise<Buffer>{
+  const jsonData = JSON.parse(this.marshal());
+  const jsonbinpackEncodedSchema = await jsonbinpack.compileSchema(${json});
+  return jsonbinpack.serialize(jsonbinpackEncodedSchema, jsonData);
+}
+
+public static async jsonbinDeserialize(buffer: Buffer): Promise<${model.name}> {
+  const jsonbinpackEncodedSchema = await jsonbinpack.compileSchema(${json});
+  const json = jsonbinpack.deserialize(jsonbinpackEncodedSchema, buffer);
+  return ${model.name}.unmarshal(json);
+}`;
+      return `${content}\n${packContent}`
+    }
+  }
+};
+
+/**
+ * Potential runtime test:
+ * 
+const instance = new Test({});
+instance.email = 'testing@test.com';
+const expectedJSON = JSON.parse(instance.marshal());
+const jsonpackbuffer = await instance.jsonbinSerialize();
+const test = await Test.jsonbinDeserialize(jsonpackbuffer);
+const actualJson = JSON.parse(test.marshal());
+expect(actualJson).toEqual(expectedJSON);
+ * 
+ */
```

Comment:
Do we need that here?

Reply:
Not necessarily, I just included it because that was what I used to validate it worked, eventually, with #612 this is the code that should be run, so did not want to loose it 😄

Reply:
So if you wanna leave it, please move to the corresponding test file :)

Reply:
Removed it 🙂